File: packages/core/src/engine.ts
import { isEssential, matchesAny, ruleAppliesTo, ruleCovers, targetDomains } from './match.js';
import type { EvalContext, Override, Policy, Rule, RuleMode, Subject } from './schema.js';
import { MINUTE, formatDuration, localTime, nextLocalMidnight, windowState } from './time.js';

/** Uso acumulado de una regla en el día. */
export interface RuleUsage {
  /** Tiempo de uso hoy (reglas de límite). */
  usedMs: number;
  /** Aperturas hoy. */
  opens: number;
  /** Tiempo de uso en el ciclo actual (reglas de intervalo). */
  intervalUsedMs: number;
  /** Fin del descanso obligatorio en curso (epoch ms). */
  breakUntil?: number;
  /** Último momento de uso registrado (epoch ms). */
  lastUsedAt?: number;
}

export interface UsageState {
  day: string;
  rules: Record<string, RuleUsage>;
}

export type BlockMode = RuleMode | 'lock';

export interface Decision {
  blocked: boolean;
  /** Regla (o 'lock') responsable del bloqueo. */
  mode?: BlockMode;
  ruleId?: string;
  ruleName?: string;
  /** Hasta cuándo dura el bloqueo; null = indefinido. */
  until?: string | null;
  /** El usuario NO puede pedir desbloqueo ni usar códigos. */
  strict: boolean;
  /** Si está permitido pero hay límites en juego: tiempo restante antes de bloquearse. */
  remainingMs?: number;
  /** Por qué se permite, si aplica. */
  allowedBy?: 'essential' | 'allowlist' | 'pause' | 'unlock';
  message: string;
}

export interface RuleStatus {
  ruleId: string;
  name: string;
  mode: RuleMode;
  enabled: boolean;
  blocked: boolean;
  until: string | null;
  usedMs: number;
  limitMs?: number;
  remainingMs?: number;
  opens?: number;
  maxOpens?: number;
  /** Próximo inicio de bloqueo programado (horarios). */
  nextBlockAt?: string;
}

export function emptyUsage(day: string): UsageState {
  return { day, rules: {} };
}

function ruleUsage(usage: UsageState, ruleId: string): RuleUsage {
  return usage.rules[ruleId] ?? { usedMs: 0, opens: 0, intervalUsedMs: 0 };
}

/** Reinicia contadores si cambió el día local. Conserva descansos en curso. */
export function ensureDay(usage: UsageState, now: Date, timeZone: string): UsageState {
  const day = localTime(now, timeZone).day;
  if (usage.day === day) return usage;
  const rules: Record<string, RuleUsage> = {};
  for (const [id, u] of Object.entries(usage.rules)) {
    if (u.breakUntil && u.breakUntil > now.getTime()) {
      rules[id] = { usedMs: 0, opens: 0, intervalUsedMs: 0, breakUntil: u.breakUntil };
    }
  }
  return { day, rules };
}

/** Suma el uso de otros dispositivos (el servidor lo envía) al uso local. */
export function mergeUsage(a: UsageState, b: UsageState): UsageState {
  if (a.day !== b.day) return a.day > b.day ? a : b;
  const rules: Record<string, RuleUsage> = { ...a.rules };
  for (const [id, u] of Object.entries(b.rules)) {
    const x = rules[id];
    if (!x) {
      rules[id] = { ...u };
      continue;
    }
    rules[id] = {
      usedMs: x.usedMs + u.usedMs,
      opens: x.opens + u.opens,
      intervalUsedMs: Math.max(x.intervalUsedMs, u.intervalUsedMs),
      breakUntil: Math.max(x.breakUntil ?? 0, u.breakUntil ?? 0) || undefined,
      lastUsedAt: Math.max(x.lastUsedAt ?? 0, u.lastUsedAt ?? 0) || undefined,
    };
  }
  return { day: a.day, rules };
}

function activeOverrides(policy: Policy, now: Date): Override[] {
  const t = now.getTime();
  return policy.overrides.filter((o) => o.type === 'bonus' || Date.parse(o.until) > t);
}

function bonusMinutes(overrides: Override[], rule: Rule, day: string): number {
  let total = 0;
  for (const o of overrides) {
    if (o.type === 'bonus' && o.day === day && (!o.ruleId || o.ruleId === rule.id)) total += o.minutes;
  }
  return total;
}

export function limitMsFor(rule: Extract<Rule, { mode: 'limit' }>, weekday: number, bonus: number): number {
  const base = rule.perDay?.[String(weekday)] ?? rule.dailyMinutes;
  return (base + bonus) * MINUTE;
}

interface RuleState {
  blocked: boolean;
  until: Date | null;
  remainingMs?: number;
  nextBlockAt?: Date;
  limitMs?: number;
}

function computeRuleState(rule: Rule, policy: Policy, usage: UsageState, now: Date, overrides: Override[]): RuleState {
  const tz = policy.timezone;
  const u = ruleUsage(usage, rule.id);
  switch (rule.mode) {
    case 'always':
      return { blocked: true, until: null };
    case 'until': {
      const until = new Date(rule.until);
      return until > now ? { blocked: true, until } : { blocked: false, until: null };
    }
    case 'schedule': {
      const ws = windowState(rule.windows, now, tz);
      const inside = ws.active;
      if (rule.invert) {
        // Permitido sólo dentro de las ventanas.
        return inside
          ? { blocked: false, until: null, remainingMs: ws.endsAt ? ws.endsAt.getTime() - now.getTime() : undefined, nextBlockAt: ws.endsAt }
          : { blocked: true, until: ws.nextStart ?? null };
      }
      return inside ? { blocked: true, until: ws.endsAt ?? null } : { blocked: false, until: null, nextBlockAt: ws.nextStart };
    }
    case 'limit': {
      const lt = localTime(now, tz);
      const limitMs = limitMsFor(rule, lt.weekday, bonusMinutes(overrides, rule, lt.day));
      const midnight = nextLocalMidnight(now, tz);
      if (rule.maxOpens !== undefined && u.opens > rule.maxOpens) return { blocked: true, until: midnight, limitMs, remainingMs: 0 };
      if (u.usedMs >= limitMs) return { blocked: true, until: midnight, limitMs, remainingMs: 0 };
      return { blocked: false, until: null, limitMs, remainingMs: limitMs - u.usedMs };
    }
    case 'interval': {
      if (rule.windows?.length && !windowState(rule.windows, now, tz).active) return { blocked: false, until: null };
      if (u.breakUntil && u.breakUntil > now.getTime()) return { blocked: true, until: new Date(u.breakUntil) };
      return { blocked: false, until: null, remainingMs: Math.max(0, rule.useMinutes * MINUTE - u.intervalUsedMs) };
    }
  }
}

function blockMessage(mode: BlockMode, name: string | undefined, until: Date | null, now: Date): string {
  const left = until ? ` Se libera en ${formatDuration(until.getTime() - now.getTime())}.` : '';
  switch (mode) {
    case 'lock':
      return `El administrador bloqueó el dispositivo.${left}`;
    case 'always':
      return `Bloqueado por “${name}”.`;
    case 'schedule':
      return `Fuera del horario permitido (“${name}”).${left}`;
    case 'limit':
      return `Se terminó el tiempo de hoy para “${name}”.${left}`;
    case 'interval':
      return `Hora de descansar (“${name}”).${left}`;
    case 'until':
      return `Bloqueo temporal activo (“${name}”).${left}`;
  }
}

function untilRank(until: Date | null): number {
  return until === null ? Number.POSITIVE_INFINITY : until.getTime();
}

function unlockCovers(o: Override, rule: Rule | null, subject: Subject): boolean {
  if (o.type !== 'unlock') return false;
  if (o.targets?.length && matchesAny(o.targets, subject)) return true;
  return !!rule && o.ruleId === rule.id;
}

/**
 * Decide si un sujeto (app o URL) debe bloquearse ahora.
 *
 * Orden de precedencia:
 *  1. Apps esenciales (llamadas, SMS, emergencias) → siempre permitidas.
 *  2. Lista permitida del admin → permitido.
 *  3. Desbloqueos temporales concedidos por el admin → permitido.
 *  4. "Liberar dispositivo" (pause) → sólo siguen activas las reglas estrictas.
 *  5. "Bloquear ya" (lock) → bloqueado.
 *  6. Reglas: si cualquiera bloquea, se bloquea (se informa la más duradera).
 *
 * Para reglas con máximo de aperturas, llamar a recordOpen() ANTES de evaluar.
 */
export function evaluate(policy: Policy, usageIn: UsageState, subject: Subject, now: Date, ctx: EvalContext = {}): Decision {
  if (isEssential(subject)) return { blocked: false, strict: false, allowedBy: 'essential', message: 'App esencial' };
  if (matchesAny(policy.allowlist, subject)) return { blocked: false, strict: false, allowedBy: 'allowlist', message: 'Permitido siempre' };

  const usage = ensureDay(usageIn, now, policy.timezone);
  const overrides = activeOverrides(policy, now);
  const paused = overrides.find((o) => o.type === 'pause');
  const lock = overrides.find((o) => o.type === 'lock');
  const subjectUnlock = overrides.find((o) => unlockCovers(o, null, subject));

  let rules = policy.rules.filter((r) => ruleCovers(r, subject, ctx) && !overrides.some((o) => unlockCovers(o, r, subject)));
  if (paused) rules = rules.filter((r) => r.strict);

  let best: { rule: Rule | null; mode: BlockMode; until: Date | null; strict: boolean } | null = null;
  let remainingMs: number | undefined;

  if (lock && !paused && !subjectUnlock) {
    best = { rule: null, mode: 'lock', until: new Date(lock.until), strict: false };
  }

  let anyStrict = false;
  for (const rule of rules) {
    const st = computeRuleState(rule, policy, usage, now, overrides);
    if (st.blocked) {
      anyStrict ||= rule.strict;
      if (!best || untilRank(st.until) > untilRank(best.until)) {
        best = { rule, mode: rule.mode, until: st.until, strict: rule.strict };
      }
    } else if (st.remainingMs !== undefined) {
      remainingMs = remainingMs === undefined ? st.remainingMs : Math.min(remainingMs, st.remainingMs);
    }
  }

  if (best) {
    return {
      blocked: true,
      mode: best.mode,
      ruleId: best.rule?.id,
      ruleName: best.rule?.name,
      until: best.until ? best.until.toISOString() : null,
      strict: best.strict || anyStrict,
      message: blockMessage(best.mode, best.rule?.name, best.until, now),
    };
  }
  if (subjectUnlock) return { blocked: false, strict: false, allowedBy: 'unlock', remainingMs, message: 'Desbloqueo temporal' };
  if (paused) return { blocked: false, strict: false, allowedBy: 'pause', message: 'Dispositivo liberado por el administrador' };
  return { blocked: false, strict: false, remainingMs, message: 'Permitido' };
}

function updateRule(usage: UsageState, ruleId: string, fn: (u: RuleUsage) => RuleUsage): UsageState {
  return { ...usage, rules: { ...usage.rules, [ruleId]: fn(ruleUsage(usage, ruleId)) } };
}

/** Registra `deltaMs` de uso en primer plano del sujeto que termina en `now`. */
export function recordUsage(policy: Policy, usageIn: UsageState, subject: Subject, deltaMs: number, now: Date, ctx: EvalContext = {}): UsageState {
  if (deltaMs <= 0 || isEssential(subject) || matchesAny(policy.allowlist, subject)) return ensureDay(usageIn, now, policy.timezone);
  let usage = ensureDay(usageIn, now, policy.timezone);
  const t = now.getTime();
  for (const rule of policy.rules) {
    if (rule.mode !== 'limit' && rule.mode !== 'interval') continue;
    if (!ruleCovers(rule, subject, ctx)) continue;
    usage = updateRule(usage, rule.id, (u) => {
      const next: RuleUsage = { ...u, usedMs: u.usedMs + deltaMs, lastUsedAt: t };
      if (rule.mode === 'interval') {
        const breakMs = rule.breakMinutes * MINUTE;
        const idleMs = u.lastUsedAt ? t - deltaMs - u.lastUsedAt : 0;
        // Si ya descansó por su cuenta el tiempo del descanso, el ciclo empieza de cero.
        const base = idleMs >= breakMs || (u.breakUntil && u.breakUntil <= t - deltaMs) ? 0 : u.intervalUsedMs;
        const used = base + deltaMs;
        if (used >= rule.useMinutes * MINUTE) {
          next.intervalUsedMs = 0;
          next.breakUntil = t + breakMs;
        } else {
          next.intervalUsedMs = used;
        }
      }
      return next;
    });
  }
  return usage;
}

/** Registra que el sujeto se abrió (para reglas con máximo de aperturas). */
export function recordOpen(policy: Policy, usageIn: UsageState, subject: Subject, now: Date, ctx: EvalContext = {}): UsageState {
  let usage = ensureDay(usageIn, now, policy.timezone);
  for (const rule of policy.rules) {
    if (rule.mode !== 'limit' || rule.maxOpens === undefined || !ruleCovers(rule, subject, ctx)) continue;
    usage = updateRule(usage, rule.id, (u) => ({ ...u, opens: u.opens + 1 }));
  }
  return usage;
}

/** Estado de cada regla para mostrar en el panel o en el agente. */
export function ruleStatuses(policy: Policy, usageIn: UsageState, now: Date, ctx: EvalContext = {}): RuleStatus[] {
  const usage = ensureDay(usageIn, now, policy.timezone);
  const overrides = activeOverrides(policy, now);
  const paused = overrides.some((o) => o.type === 'pause');
  return policy.rules
    .filter((r) => ruleAppliesTo(r, ctx))
    .map((rule) => {
      const u = ruleUsage(usage, rule.id);
      const st = computeRuleState(rule, policy, usage, now, overrides);
      const unlocked = overrides.some((o) => o.type === 'unlock' && o.ruleId === rule.id);
      const blocked = rule.enabled && st.blocked && !unlocked && (!paused || rule.strict);
      return {
        ruleId: rule.id,
        name: rule.name,
        mode: rule.mode,
        enabled: rule.enabled,
        blocked,
        until: blocked && st.until ? st.until.toISOString() : null,
        usedMs: u.usedMs,
        limitMs: st.limitMs,
        remainingMs: st.remainingMs,
        opens: rule.mode === 'limit' && rule.maxOpens !== undefined ? u.opens : undefined,
        maxOpens: rule.mode === 'limit' ? rule.maxOpens : undefined,
        nextBlockAt: st.nextBlockAt?.toISOString(),
      };
    });
}

export interface WebBlockList {
  /** Bloquear todo el tráfico web salvo `allowedDomains`. */
  blockAll: boolean;
  blockedDomains: string[];
  allowedDomains: string[];
}

/**
 * Lista plana de dominios bloqueados AHORA, para mecanismos que sólo entienden
 * dominios (archivo hosts, DNS). Las reglas de URL/palabra clave requieren un
 * agente con visibilidad de URL (extensión de navegador, accesibilidad Android).
 */
export function currentWebBlockList(policy: Policy, usageIn: UsageState, now: Date, ctx: EvalContext = {}): WebBlockList {
  const usage = ensureDay(usageIn, now, policy.timezone);
  const overrides = activeOverrides(policy, now);
  const paused = overrides.some((o) => o.type === 'pause');
  const lock = !paused && overrides.some((o) => o.type === 'lock');
  const blocked = new Set<string>();
  const allowed = new Set<string>(policy.allowlist.flatMap(targetDomains));
  for (const o of overrides) if (o.type === 'unlock') o.targets?.flatMap(targetDomains).forEach((d) => allowed.add(d));
  let blockAll = lock;

  for (const rule of policy.rules) {
    if (!rule.enabled || !ruleAppliesTo(rule, ctx) || (paused && !rule.strict)) continue;
    if (overrides.some((o) => o.type === 'unlock' && o.ruleId === rule.id)) continue;
    if (!computeRuleState(rule, policy, usage, now, overrides).blocked) continue;
    const except = new Set(rule.exceptions.flatMap(targetDomains));
    if (rule.targets.some((t) => t.kind === 'all')) {
      blockAll = true;
      except.forEach((d) => allowed.add(d));
    }
    for (const d of rule.targets.flatMap(targetDomains)) if (!except.has(d)) blocked.add(d);
  }
  for (const d of allowed) blocked.delete(d);
  return { blockAll, blockedDomains: [...blocked].sort(), allowedDomains: [...allowed].sort() };
}

File: apps/android/app/src/main/java/com/guardian/blocker/engine/Engine.kt
package com.guardian.blocker.engine

import java.net.URLDecoder
import java.time.Instant
import java.time.ZoneId
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min

/*
 * Port a Kotlin del motor de packages/core (time.ts, match.ts, engine.ts).
 * Debe comportarse igual: las pruebas de EngineTest replican las de TypeScript.
 */

const val MINUTE = 60_000L
private const val DAY_MINUTES = 24 * 60

// ------------------------------------------------------------------ hora local

data class LocalTime(val day: String, val weekday: Int, val minutes: Double)

fun localTime(nowMs: Long, timezone: String): LocalTime {
    val zone = runCatching { ZoneId.of(timezone) }.getOrDefault(ZoneId.systemDefault())
    val z = Instant.ofEpochMilli(nowMs).atZone(zone)
    return LocalTime(
        day = z.toLocalDate().toString(),
        weekday = z.dayOfWeek.value % 7, // domingo = 0
        minutes = z.hour * 60 + z.minute + z.second / 60.0,
    )
}

fun nextLocalMidnight(nowMs: Long, timezone: String): Long =
    nowMs + ((DAY_MINUTES - localTime(nowMs, timezone).minutes) * MINUTE).toLong()

fun parseHHMM(v: String): Int {
    val (h, m) = v.split(":").map { it.toInt() }
    return h * 60 + m
}

private fun minutesLeftInWindow(w: TimeWindow, lt: LocalTime): Double? {
    val s = parseHHMM(w.start)
    val e = parseHHMM(w.end)
    val today = lt.weekday in w.days
    val yesterday = ((lt.weekday + 6) % 7) in w.days
    if (e > s) return if (today && lt.minutes >= s && lt.minutes < e) e - lt.minutes else null
    if (today && lt.minutes >= s) return DAY_MINUTES - lt.minutes + e
    if (yesterday && lt.minutes < e) return e - lt.minutes
    return null
}

data class WindowState(val active: Boolean, val endsAt: Long? = null, val nextStart: Long? = null)

fun windowState(windows: List<TimeWindow>, nowMs: Long, tz: String): WindowState {
    var cursor = nowMs
    var active = false
    repeat(8) {
        val lt = localTime(cursor, tz)
        val best = windows.mapNotNull { minutesLeftInWindow(it, lt) }.maxOrNull() ?: return@repeat
        active = true
        cursor += ceil(best * MINUTE).toLong()
    }
    if (active) return WindowState(true, endsAt = cursor)
    val lt = localTime(nowMs, tz)
    var best: Double? = null
    for (offset in 0..7) {
        val wd = (lt.weekday + offset) % 7
        for (w in windows) {
            if (wd !in w.days) continue
            val delta = offset * DAY_MINUTES + parseHHMM(w.start) - lt.minutes
            if (delta > 0 && (best == null || delta < best)) best = delta
        }
    }
    return WindowState(false, nextStart = best?.let { nowMs + ceil(it * MINUTE).toLong() })
}

fun formatDuration(ms: Long): String {
    val total = max(0, Math.round(ms / MINUTE.toDouble()))
    val h = total / 60
    val m = total % 60
    return when {
        h == 0L -> "$m min"
        m == 0L -> "$h h"
        else -> "$h h $m min"
    }
}

// ------------------------------------------------------------------ coincidencias

fun normalizeHost(input: String): String {
    var s = input.trim().lowercase()
    s = s.replace(Regex("^[a-z][a-z0-9+.-]*://"), "")
    s = s.split(Regex("[/?#]"))[0]
    s = s.replace(Regex("^[^@]*@"), "").replace(Regex(":\\d+$"), "").removeSuffix(".")
    s = s.removePrefix("*.").removePrefix("www.")
    return s
}

fun normalizeUrl(input: String): String {
    val s = input.trim().replace(Regex("^[a-zA-Z][a-zA-Z0-9+.-]*://"), "")
    val slash = s.indexOfFirst { it == '/' || it == '?' || it == '#' }
    val host = normalizeHost(if (slash == -1) s else s.substring(0, slash))
    val rest = if (slash == -1) "" else s.substring(slash)
    val decoded = runCatching { URLDecoder.decode(rest.replace("+", "%2B"), "UTF-8") }.getOrDefault(rest)
    return (host + decoded).lowercase()
}

fun domainMatches(host: String, domain: String): Boolean {
    val h = normalizeHost(host)
    val d = normalizeHost(domain)
    return d.isNotEmpty() && (h == d || h.endsWith(".$d"))
}

private fun urlPrefixMatches(url: String, prefix: String): Boolean {
    val u = normalizeUrl(url)
    val p = normalizeUrl(prefix)
    val ps = p.indexOf('/')
    val pHost = if (ps == -1) p else p.substring(0, ps)
    val pPath = if (ps == -1) "" else p.substring(ps)
    val us = u.indexOfFirst { it == '/' || it == '?' || it == '#' }
    val uHost = if (us == -1) u else u.substring(0, us)
    val uPath = if (us == -1) "" else u.substring(us)
    return domainMatches(uHost, pHost) && uPath.startsWith(pPath)
}

class Matcher(private val catalog: Catalog) {
    fun isEssential(subject: Subject) = subject is Subject.App && subject.id in catalog.essentialApps

    fun matches(t: Target, s: Subject): Boolean = when (t) {
        Target.All -> true
        is Target.App -> s is Subject.App && (t.platform == null || t.platform == s.platform) && t.id == s.id
        is Target.Service -> catalog.service(t.id)?.let { svc ->
            when (s) {
                is Subject.App -> s.id in svc.apps
                is Subject.Web -> svc.domains.any { domainMatches(s.url, it) }
            }
        } ?: false
        is Target.Domain -> s is Subject.Web && domainMatches(s.url, t.domain)
        is Target.Url -> s is Subject.Web && urlPrefixMatches(s.url, t.prefix)
        is Target.Keyword -> s is Subject.Web && normalizeUrl(s.url).contains(t.keyword.trim().lowercase())
        is Target.Category -> catalog.inCategory(t.category).any { matches(Target.Service(it.id), s) } ||
            (s is Subject.Web && (
                catalog.categoryDomains[t.category].orEmpty().any { domainMatches(s.url, it) } ||
                    catalog.categoryKeywords[t.category].orEmpty().any { normalizeHost(s.url).contains(it) }
                ))
        is Target.IosSelection -> false
    }

    fun matchesAny(targets: List<Target>, s: Subject) = targets.any { matches(it, s) }

    fun appliesTo(rule: Rule, platform: String?, deviceId: String?): Boolean {
        if (!rule.platforms.isNullOrEmpty() && platform != null && platform !in rule.platforms) return false
        if (!rule.deviceIds.isNullOrEmpty() && deviceId != null && deviceId !in rule.deviceIds) return false
        return true
    }

    fun covers(rule: Rule, s: Subject, platform: String?, deviceId: String?) =
        rule.enabled && appliesTo(rule, platform, deviceId) && matchesAny(rule.targets, s) && !matchesAny(rule.exceptions, s)

    fun domainsOf(t: Target): List<String> = when (t) {
        is Target.Service -> catalog.service(t.id)?.domains.orEmpty()
        is Target.Domain -> listOf(normalizeHost(t.domain))
        is Target.Category -> catalog.inCategory(t.category).flatMap { it.domains } + catalog.categoryDomains[t.category].orEmpty()
        else -> emptyList()
    }

    fun appIdsOf(t: Target, platform: String): List<String> = when (t) {
        is Target.App -> if (t.platform == null || t.platform == platform) listOf(t.id) else emptyList()
        is Target.Service -> catalog.service(t.id)?.apps.orEmpty()
        is Target.Category -> catalog.inCategory(t.category).flatMap { it.apps }
        else -> emptyList()
    }
}

// ------------------------------------------------------------------ uso

data class RuleUsage(
    val usedMs: Long = 0,
    val opens: Int = 0,
    val intervalUsedMs: Long = 0,
    val breakUntil: Long? = null,
    val lastUsedAt: Long? = null,
)

data class UsageState(val day: String, val rules: Map<String, RuleUsage> = emptyMap()) {
    fun rule(id: String) = rules[id] ?: RuleUsage()
    fun with(id: String, u: RuleUsage) = copy(rules = rules + (id to u))
}

fun ensureDay(u: UsageState, nowMs: Long, tz: String): UsageState {
    val day = localTime(nowMs, tz).day
    if (u.day == day) return u
    return UsageState(day, u.rules.filter { (it.value.breakUntil ?: 0) > nowMs }.mapValues { RuleUsage(breakUntil = it.value.breakUntil) })
}

fun mergeUsage(a: UsageState, b: UsageState): UsageState {
    if (a.day != b.day) return if (a.day > b.day) a else b
    val out = a.rules.toMutableMap()
    for ((id, u) in b.rules) {
        val x = out[id]
        out[id] = if (x == null) u else RuleUsage(
            usedMs = x.usedMs + u.usedMs,
            opens = x.opens + u.opens,
            intervalUsedMs = max(x.intervalUsedMs, u.intervalUsedMs),
            breakUntil = listOfNotNull(x.breakUntil, u.breakUntil).maxOrNull(),
            lastUsedAt = listOfNotNull(x.lastUsedAt, u.lastUsedAt).maxOrNull(),
        )
    }
    return UsageState(a.day, out)
}

// ------------------------------------------------------------------ motor

data class RuleStatus(
    val rule: Rule,
    val blocked: Boolean,
    val until: Long?,
    val usedMs: Long,
    val limitMs: Long?,
    val remainingMs: Long?,
    val nextBlockAt: Long?,
)

class Engine(private val catalog: Catalog, private val platform: String, private val deviceId: String?) {
    val matcher = Matcher(catalog)

    private data class State(val blocked: Boolean, val until: Long?, val remainingMs: Long? = null, val nextBlockAt: Long? = null, val limitMs: Long? = null)

    private fun activeOverrides(p: Policy, now: Long) = p.overrides.filter { it.type == "bonus" || (it.until ?: 0) > now }

    private fun bonus(overrides: List<Override>, rule: Rule, day: String) =
        overrides.filter { it.type == "bonus" && it.day == day && (it.ruleId == null || it.ruleId == rule.id) }.sumOf { it.minutes }

    private fun state(rule: Rule, p: Policy, usage: UsageState, now: Long, overrides: List<Override>): State {
        val tz = p.timezone
        val u = usage.rule(rule.id)
        return when (rule.mode) {
            "always" -> State(true, null)
            "until" -> if ((rule.until ?: 0) > now) State(true, rule.until) else State(false, null)
            "schedule" -> {
                val ws = windowState(rule.windows, now, tz)
                if (rule.invert) {
                    if (ws.active) State(false, null, remainingMs = ws.endsAt?.minus(now), nextBlockAt = ws.endsAt) else State(true, ws.nextStart)
                } else {
                    if (ws.active) State(true, ws.endsAt) else State(false, null, nextBlockAt = ws.nextStart)
                }
            }
            "limit" -> {
                val lt = localTime(now, tz)
                val limitMs = ((rule.perDay[lt.weekday] ?: rule.dailyMinutes) + bonus(overrides, rule, lt.day)) * MINUTE
                val midnight = nextLocalMidnight(now, tz)
                when {
                    rule.maxOpens != null && u.opens > rule.maxOpens -> State(true, midnight, 0, limitMs = limitMs)
                    u.usedMs >= limitMs -> State(true, midnight, 0, limitMs = limitMs)
                    else -> State(false, null, limitMs - u.usedMs, limitMs = limitMs)
                }
            }
            "interval" -> when {
                rule.windows.isNotEmpty() && !windowState(rule.windows, now, tz).active -> State(false, null)
                (u.breakUntil ?: 0) > now -> State(true, u.breakUntil)
                else -> State(false, null, max(0, rule.useMinutes * MINUTE - u.intervalUsedMs))
            }
            else -> State(false, null)
        }
    }

    private fun message(mode: String, name: String?, until: Long?, now: Long): String {
        val left = until?.let { " Se libera en ${formatDuration(it - now)}." } ?: ""
        return when (mode) {
            "lock" -> "El administrador bloqueó el dispositivo.$left"
            "always" -> "Bloqueado por “$name”."
            "schedule" -> "Fuera del horario permitido (“$name”).$left"
            "limit" -> "Se terminó el tiempo de hoy para “$name”.$left"
            "interval" -> "Hora de descansar (“$name”).$left"
            else -> "Bloqueo temporal activo (“$name”).$left"
        }
    }

    private fun unlockCovers(o: Override, rule: Rule?, s: Subject): Boolean {
        if (o.type != "unlock") return false
        if (!o.targets.isNullOrEmpty() && matcher.matchesAny(o.targets, s)) return true
        return rule != null && o.ruleId == rule.id
    }

    fun evaluate(p: Policy, usageIn: UsageState, s: Subject, now: Long): Decision {
        if (matcher.isEssential(s)) return Decision(false, allowedBy = "essential", message = "App esencial")
        if (matcher.matchesAny(p.allowlist, s)) return Decision(false, allowedBy = "allowlist", message = "Permitido siempre")
        val usage = ensureDay(usageIn, now, p.timezone)
        val overrides = activeOverrides(p, now)
        val paused = overrides.firstOrNull { it.type == "pause" }
        val lock = overrides.firstOrNull { it.type == "lock" }
        val subjectUnlock = overrides.firstOrNull { unlockCovers(it, null, s) }

        var rules = p.rules.filter { r -> matcher.covers(r, s, platform, deviceId) && overrides.none { unlockCovers(it, r, s) } }
        if (paused != null) rules = rules.filter { it.strict }

        var bestRule: Rule? = null
        var bestMode: String? = null
        var bestUntil: Long? = null
        var bestSet = false
        var anyStrict = false
        var remaining: Long? = null
        fun rank(u: Long?) = u ?: Long.MAX_VALUE

        if (lock != null && paused == null && subjectUnlock == null) {
            bestMode = "lock"; bestUntil = lock.until; bestSet = true
        }
        for (rule in rules) {
            val st = state(rule, p, usage, now, overrides)
            if (st.blocked) {
                anyStrict = anyStrict || rule.strict
                if (!bestSet || rank(st.until) > rank(bestUntil)) {
                    bestRule = rule; bestMode = rule.mode; bestUntil = st.until; bestSet = true
                }
            } else if (st.remainingMs != null) {
                remaining = remaining?.let { min(it, st.remainingMs) } ?: st.remainingMs
            }
        }
        if (bestSet) {
            return Decision(
                blocked = true,
                mode = bestMode,
                ruleId = bestRule?.id,
                ruleName = bestRule?.name,
                until = bestUntil,
                strict = (bestRule?.strict ?: false) || anyStrict,
                message = message(bestMode!!, bestRule?.name, bestUntil, now),
            )
        }
        if (subjectUnlock != null) return Decision(false, remainingMs = remaining, allowedBy = "unlock", message = "Desbloqueo temporal")
        if (paused != null) return Decision(false, allowedBy = "pause", message = "Dispositivo liberado por el administrador")
        return Decision(false, remainingMs = remaining, message = "Permitido")
    }

    fun recordUsage(p: Policy, usageIn: UsageState, s: Subject, deltaMs: Long, now: Long): UsageState {
        var usage = ensureDay(usageIn, now, p.timezone)
        if (deltaMs <= 0 || matcher.isEssential(s) || matcher.matchesAny(p.allowlist, s)) return usage
        for (rule in p.rules) {
            if (rule.mode != "limit" && rule.mode != "interval") continue
            if (!matcher.covers(rule, s, platform, deviceId)) continue
            val u = usage.rule(rule.id)
            var next = u.copy(usedMs = u.usedMs + deltaMs, lastUsedAt = now)
            if (rule.mode == "interval") {
                val breakMs = rule.breakMinutes * MINUTE
                val idle = u.lastUsedAt?.let { now - deltaMs - it } ?: 0
                val reset = idle >= breakMs || (u.breakUntil != null && u.breakUntil <= now - deltaMs)
                val used = (if (reset) 0 else u.intervalUsedMs) + deltaMs
                next = if (used >= rule.useMinutes * MINUTE) next.copy(intervalUsedMs = 0, breakUntil = now + breakMs) else next.copy(intervalUsedMs = used)
            }
            usage = usage.with(rule.id, next)
        }
        return usage
    }

    fun recordOpen(p: Policy, usageIn: UsageState, s: Subject, now: Long): UsageState {
        var usage = ensureDay(usageIn, now, p.timezone)
        for (rule in p.rules) {
            if (rule.mode != "limit" || rule.maxOpens == null || !matcher.covers(rule, s, platform, deviceId)) continue
            val u = usage.rule(rule.id)
            usage = usage.with(rule.id, u.copy(opens = u.opens + 1))
        }
        return usage
    }

    fun statuses(p: Policy, usageIn: UsageState, now: Long): List<RuleStatus> {
        val usage = ensureDay(usageIn, now, p.timezone)
        val overrides = activeOverrides(p, now)
        val paused = overrides.any { it.type == "pause" }
        return p.rules.filter { matcher.appliesTo(it, platform, deviceId) }.map { rule ->
            val st = state(rule, p, usage, now, overrides)
            val unlocked = overrides.any { it.type == "unlock" && it.ruleId == rule.id }
            val blocked = rule.enabled && st.blocked && !unlocked && (!paused || rule.strict)
            RuleStatus(rule, blocked, if (blocked) st.until else null, usage.rule(rule.id).usedMs, st.limitMs, st.remainingMs, st.nextBlockAt)
        }
    }

    /** Apps concretas bloqueadas ahora (para suspenderlas si somos propietarios del dispositivo). */
    fun blockedPackagesNow(p: Policy, usage: UsageState, installed: Collection<String>, now: Long): Set<String> =
        installed.filter { pkg ->
            val s = Subject.App(platform, pkg)
            // Sólo reglas con objetivos concretos: "todo el dispositivo" lo cubre la pantalla de bloqueo.
            p.rules.any { r -> r.targets.any { it !is Target.All && matcher.matches(it, s) } } && evaluate(p, usage, s, now).blocked
        }.toSet()
}
